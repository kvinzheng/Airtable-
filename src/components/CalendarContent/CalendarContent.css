.CalendarContent {
  display: grid;
  grid-template-columns: repeat(35, 1fr);
  position: relative;
  overflow-x: scroll;
  overflow-y: hidden;
}
.CalendarContent-week {
  display: flex;
  justify-content: center;
  text-transform: uppercase;
  font-size: 0.9rem;
  text-align: center;
  padding: 10px 0px;
  border-bottom: 1px solid #dfdfdf;
}

.CalendarContent-outlined {
  background: #ffffe5;
}

.CalendarContent-extra {
  background: #f7f7f7;
  color: rgb(201, 201, 201);
}
