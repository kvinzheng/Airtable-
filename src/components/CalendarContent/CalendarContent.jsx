import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { DragDropContext } from "react-beautiful-dnd";

import { daysDifference } from "../../helpers/eventTimeline";
import CalendarCell from "../CalendarCell/CalendarCell";
import { setEvents } from "../../actions/events";
import {
  weekdays,
  getArrayOfDates,
  doMonthsMatch,
  istheSameDay,
  makeDateOfDateAndTime,
} from "../../helpers/dateFormat";

import "./CalendarContent.css";

export const CalendarContent = ({
  displayedDate,
  events,
  outlinedDate,
  eventPosition,
  zoomWidth,
  setEvents,
}) => {
  const handleEventDrag = (result) => {
    const { destination, source, draggableId } = result;

    if (source.droppableId !== destination.droppableId) {
      const newEvents = events.map((event) => {
        const newStartDate = makeDateOfDateAndTime(
          destination.droppableId,
          event.date
        );
        const newEndDate = new Date(event.end.getTime());
        const diff = daysDifference(event.date, event.end);

        newEndDate.setDate(newStartDate.getDate() + diff);

        const newEvent = {
          ...event,
          date: newStartDate,
          end: newEndDate,
        };

        return event.id === draggableId ? newEvent : event;
      });
      setEvents(newEvents);
    }
  };

  const onDrag = (result) => {
    const { destination, source } = result;
    const inValid =
      !destination ||
      (destination.droppableId === source.droppableId &&
        destination.index === source.index);

    if (inValid) return;
    handleEventDrag(result);
  };

  const getCellType = (date) => {
    if (istheSameDay(date, outlinedDate)) {
      return "CalendarContent-outlined";
    }

    if (!doMonthsMatch(date, displayedDate)) {
      return "CalendarContent-extra";
    }
    return "";
  };

  const dates = getArrayOfDates(displayedDate);
  const lastDate = dates[dates.length - 1];

  const monthlyWeekdays = [];
  for (let i = 0; i < 5; i++) {
    monthlyWeekdays.push(...weekdays);
  }

  return (
    <div className="CalendarContent">
      {monthlyWeekdays.map((weekday, index) => (
        <div className="CalendarContent-week" tabIndex={-1} key={index}>
          <div style={{ width: `${zoomWidth}px` }}> {weekday}</div>
        </div>
      ))}

      <DragDropContext  onDragEnd={onDrag}>
        {dates.map((date, index) => {
          return (
            <CalendarCell
              key={index}
              date={date}
              events={events}
              cellIndex={date.valueOf()}
              type={getCellType(date)}
              eventPosition={eventPosition}
              lastDate={lastDate}
            />
          );
        })}
      </DragDropContext>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    events: state.events.data,
    form: state.form.calendarForm,
    zoomWidth: state.events.zoom.width,
  };
};

CalendarContent.propTypes = {
  events: PropTypes.array,
  form: PropTypes.bool,
  zoomWidth: PropTypes.number,
};

export default connect(mapStateToProps, { setEvents })(CalendarContent);
