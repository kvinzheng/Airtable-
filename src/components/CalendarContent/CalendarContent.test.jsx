import React from "react";

import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { CalendarContent } from "./CalendarContent";
import MockData from "../../helpers/testingData";

configure({ adapter: new Adapter() });

describe("CalendarContent", () => {
  it("should render without any data defined", () => {
    const component = shallow(<CalendarContent />);

    expect(component).toMatchSnapshot();
  });

  it("should render with data defined", () => {
    const component = shallow(
      <CalendarContent
        events={MockData.events.data}
        form={MockData.form.calendarForm}
        zoomWidth={MockData.events.zoom.width}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
