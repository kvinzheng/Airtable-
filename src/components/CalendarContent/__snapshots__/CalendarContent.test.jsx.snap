// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarContent should render with data defined 1`] = `ShallowWrapper {}`;

exports[`CalendarContent should render without any data defined 1`] = `ShallowWrapper {}`;
